packages/ra-core/src/controller/input/useReferenceArrayInputController.ts

```diff
@@ -239,57 +111,47 @@ export const useReferenceArrayInputController = <
             ? finalReferenceRecords
             : matchingReferences;
 
-    const dataStatus = getDataStatus<RecordType>({
-        field,
-        matchingReferences: finalMatchingReferences,
-        referenceRecords: finalReferenceRecords,
-        translate,
-    });
-
     const refetch = useCallback(() => {
         refetchGetMany();
         refetchGetMatching();
     }, [refetchGetMany, refetchGetMatching]);
 
+    const currentSort = useMemo(
+        () => ({
+            field: params.sort,
+            order: params.order,
+        }),
+        [params.sort, params.order]
+    );
     return {
-        displayedFilters,
-        choices: dataStatus.choices,
-        warning: dataStatus.warning,
-        error:
-            errorGetMany || errorGetList
-                ? translate('ra.input.references.all_missing', {
-                      _: 'ra.input.references.all_missing',
-                  })
-                : undefined,
+        sort: currentSort,
+        allChoices: finalMatchingReferences,
+        availableChoices: matchingReferences,
+        selectedChoices: finalReferenceRecords,
+        displayedFilters: params.displayedFilters,
+        error: errorGetMany || errorGetList,
+        filter,
+        filterValues: params.filterValues,
+        hideFilter: paramsModifiers.hideFilter,
         isFetching: isFetchingGetMany || isFetchingGetList,
         isLoading: isLoadingGetMany || isLoadingGetList,
-        // possibleValues
-        data: matchingReferences,
-        total,
-        sort,
-        setSort,
-        filterValues,
-        hideFilter,
-        showFilter,
-        setFilter,
-        setFilters,
-        onSelect,
-        onToggleItem,
-        onUnselectItems,
-        page,
-        perPage,
-        setPage,
-        setPagination,
-        setPerPage,
+        page: params.page,
+        perPage: params.perPage,
+        refetch,
+        resource: reference,
+        setFilters: paramsModifiers.setFilters,
+        setPage: paramsModifiers.setPage,
+        setPerPage: paramsModifiers.setPerPage,
+        setSort: paramsModifiers.setSort,
+        showFilter: paramsModifiers.showFilter,
+        source,
+        total: total,
         hasNextPage: pageInfo
             ? pageInfo.hasNextPage
             : total != null
-            ? page * perPage < total
+            ? params.page * params.perPage < total
             : undefined,
-        hasPreviousPage: pageInfo ? pageInfo.hasPreviousPage : page > 1,
-        refetch,
-        resource,
-        selectedIds: field.value || EmptyArray,
```

Comment:
I understand the controller no longer manages a "selection", and that it becomes the responsibility of its choild (e.g. DatagridInput) to to it. I think this should be mentioned in the upgrade guide.